### PR TITLE
Taxonomie with tags devas 1493

### PR DESCRIPTION
**Pull request by amensouissi**

### CODE REVIEW

File: assembl/static2/js/app/components/form/selectFieldAdapter.jsx [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+  placeholder,
+  className,
+  options,
+  loadOptions,
+  input: { name, onChange, value, ...otherListeners },
+  meta: { error, touched },
+  ...rest
+}: Props) => {
+  const decoratedLabel = label && required ? `${label} *` : label;
+  let SelectComponent = null;
```
[the remark this answers is not in the document]
Comment (amensouissi) [reply assembl/static2/js/app/components/form/selectFieldAdapter.jsx]: it is just the initial value. SelectComponent must be an asynchronous or a synchronous Select component

File: assembl/static2/js/app/components/harvesting/tagsForm.jsx [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+    .query({
+      query: Tags,
+      variables: { filter: inputValue },
+      fetchPolicy: 'network-only'
+    })
+    .then((value) => {
+      const { data: { tags } } = value;
+      return tags.map(tag => ({
+        label: tag.value,
+        value: tag.id
```
Comment (MilanBarande) [inline assembl/static2/js/app/components/harvesting/tagsForm.jsx]: Why not calling it id?

Comment (amensouissi) [reply assembl/static2/js/app/components/harvesting/tagsForm.jsx]: it's the react-select convention

**Comment by amensouissi** [comment]
TODO

- [ ] Fix test
- [ ] Fix menu position

File: assembl/graphql/extract.py [hunk 3]
```diff
@@ -121,6 +147,8 @@ def mutate(root, args, context, info):
         require_instance_permission(CrudPermissions.DELETE, extract, context)
         for fragment in extract.text_fragment_identifiers:
             fragment.delete()
+
+        extract.tags = []
```
Comment (cedrtang) [inline assembl/graphql/extract.py]: Is it necessary to clear the tags before deleting the Extract ? Is the extract still stored in db if deleted ?

Comment (cedrtang) [reply assembl/graphql/extract.py]: Also, if the last extract that used a specific tag is deleted, what happens to the tag not used anymore ?

File: assembl/static2/js/app/utils/translations.js [hunk 4]
```diff
@@ -56,7 +56,21 @@ const Translations = {
...
+      tags: {
+        label: "Tags",
+        cancel: "Annuler",
+        validate: "Valider",
+        deleteConfirmation: {
+          confirmDeletionTitle: "Supprimer le tag \"%{tag}\"",
+          confirmDeletionBody: "Etes-vous certain de vouloir supprimer ce tag ?"
+        },
+        select: {
+          placeholder: "Selectionner des tags",
```
Comment (cedrtang) [inline assembl/static2/js/app/utils/translations.js]: Just missing 2 accents on "Êtes" and "sélectionner"

File: assembl/static2/js/app/components/harvesting/tagForm.jsx [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+
+type Props = {
+  client: ApolloClient,
+  initialValue: Option,
+  excludeOptions: Array<string>,
+  onSubmit: (result: { tags: Array<Option> }) => void,
+  onCancel: () => void
+};
+
+const valueContainer = (pristine: boolean, onValidate?: Function, onCancel?: Function) => ({ children, ...props }) => {
```
Comment (cedrtang) [inline assembl/static2/js/app/components/harvesting/tagForm.jsx]: Would it be valuable to define the format of onValidate and onCancel here too ?

File: assembl/static2/js/app/components/harvesting/harvestingBox.jsx [hunk 6]
```diff
@@ -102,7 +150,7 @@ class DumbHarvestingBox extends React.Component<Props, State> {
       [ACTIONS.create]: {
         buttons: [
           { id: 'cancel', title: 'debate.confirmDeletionButtonCancel', className: 'button-cancel', onClick: cancelHarvesting },
-          { id: 'validate', title: 'harvesting.submit', className: 'button-submit', onClick: this.validateHarvesting }
+          { id: 'validate', title: 'harvesting.submit', className: 'button-submit', onClick: null }
```
[the remark this answers is not in the document]
Comment (amensouissi) [reply assembl/static2/js/app/components/harvesting/harvestingBox.jsx]: yes the handleSubmit of the react-final-form

File: assembl/graphql/extract.py (hunk 3 above)
Comment (amensouissi) [reply assembl/graphql/extract.py]: yes, as with text_fragment_identifiers, we need to remove tag associations and not tags.

File: assembl/graphql/schema.py [hunk 7]
```diff
@@ -133,6 +135,24 @@ class Query(graphene.ObjectType):
...
+        description=docs.SchemaTags.__doc__)
+
+    def resolve_tags(self, args, context, info):
+        discussion_id = context.matchdict['discussion_id']
+        _filter = args.get('filter', '')
+        model = models.Keyword
+        query = get_query(model, context).filter(
+            model.discussion_id == discussion_id)
+
+        # @TODO Add Pagination
```
Comment (vincentfretin) [inline assembl/graphql/schema.py]: If you want pagination, transform List to a SQLAlchemyConnectionField and remove the `.all()` in the query below.

File: assembl/graphql/tag.py [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+    value = graphene.String(description=docs.TagInterface.value)
+
+
+class Tag(SecureObjectType, SQLAlchemyObjectType):
+    __doc__ = docs.TagInterface.__doc__
+
+    class Meta:
+        model = models.Keyword
+        interfaces = (Node, TagInterface)
+        only_fields = ('id')
```
Comment (vincentfretin) [inline assembl/graphql/tag.py]: Oh, does it work? Shouldn't it be `('id', )`?

File: assembl/static2/js/app/components/common/tooltips.jsx [hunk 9]
```diff
@@ -216,6 +216,12 @@ export const harvestingTooltip = (
   </Tooltip>
 );
 
+export const eidtTagTooltip = (
```
Comment (vincentfretin) [inline assembl/static2/js/app/components/common/tooltips.jsx]: eidtTagTooltip => editTagTooltip

File: assembl/static2/js/app/components/common/tooltips.jsx [hunk 10]
```diff
@@ -216,6 +216,12 @@ export const harvestingTooltip = (
...
+  <Tooltip id="harvestingTooltip">
```
Comment (vincentfretin) [inline assembl/static2/js/app/components/common/tooltips.jsx]: This id is already used, change it.

File: assembl/static2/js/app/components/harvesting/harvestingBox.jsx [hunk 11]
```diff
@@ -61,12 +67,54 @@ type Taxonomies = {
...
+      },
+      optimisticResponse: {
+        __typename: 'Mutation',
+        updateExtractTags: {
+          __typename: 'UpdateExtractTags',
+          tags: tags.map(tag => ({ __typename: 'Tag', value: tag, id: tag }))
+        }
+      },
+      updateQueries: {
+        // Udate the Post query
```
Comment (vincentfretin) [inline assembl/static2/js/app/components/harvesting/harvestingBox.jsx]: Udate => Update

File: assembl/static2/js/app/components/harvesting/harvestingBox.jsx [hunk 12]
```diff
@@ -367,6 +416,24 @@ class DumbHarvestingBox extends React.Component<Props, State> {
     }));
   };
 
+  updateTags = (tags: Array<string>, callback: () => void) => {
+    const { extractIndex } = this.state;
+    const extract = this.getCurrentExtract(extractIndex);
```
Comment (vincentfretin) [inline assembl/static2/js/app/components/harvesting/harvestingBox.jsx]: You will need to use getCurrentExtractByIndex after #612 is merged

File: assembl/static2/js/app/components/harvesting/harvestingBox.jsx [hunk 13]
```diff
@@ -375,6 +442,23 @@ class DumbHarvestingBox extends React.Component<Props, State> {
     const actionId = isEditable ? ACTIONS.edit : (disabled && ACTIONS.create) || (isSubmitted && ACTIONS.confirm);
     if (!actionId) return null;
     const action = this.actions[actionId];
+    if (disabled) {
+      // Render the Tags form ony if it is the create action
```
Comment (vincentfretin) [inline assembl/static2/js/app/components/harvesting/harvestingBox.jsx]: ony => only

File: assembl/static2/js/app/components/harvesting/tag.jsx [hunk 14]
```diff
@@ -0,0 +1,173 @@
...
+      },
+      optimisticResponse: {
+        __typename: 'Mutation',
+        updateTag: {
+          __typename: 'UpdateTag',
+          tag: { __typename: 'Tag', value: value, id: value }
+        }
+      },
+      updateQueries: {
+        // Udate the Post query
```
Comment (vincentfretin) [inline assembl/static2/js/app/components/harvesting/tag.jsx]: Udate => Update

File: assembl/tests/fixtures/idea_content_links.py [hunk 15]
```diff
@@ -133,11 +133,19 @@ def extract_post_1_to_subidea_1_1(
...
     test_session.flush()
 
     def fin():
         print "finalizer extract_post_1_to_subidea_1_1"
+        tags = e.tags
+        e.tags = []
+        for tag in tags:
+            test_session.delete(tag)
+
+        test_session.delete(e)
```
Comment (vincentfretin) [inline assembl/tests/fixtures/idea_content_links.py]: You have `test_session.delete(e)` twice.

File: assembl/graphql/tag.py [hunk 16]
```diff
@@ -0,0 +1,82 @@
...
+class TagInterface(SQLAlchemyInterface):
+    __doc__ = docs.TagInterface.__doc__
+
+    class Meta:
+        model = models.Keyword
+        only_fields = ('value', )
+        # Don't add id in only_fields in an interface
+        # will be just the primary key, not the base64 type:id
+
+    value = graphene.String(description=docs.TagInterface.value)
```
Comment (vincentfretin) [inline assembl/graphql/tag.py]: You should maybe add `required=True`?
You don't need to list `value` in `only_fields` if you write `value = ...`.

**Pull request merged**